File: api/app/services/password_policy.py
"""
Password complexity rules for account creation. Deterministic, no LLM
involvement (§41's "deterministic before LLM" applies just as much to a
security rule as to a stats parser) — every rule here is checkable by plain
Python, and the exact same rule set is mirrored client-side in
app/utils/passwordPolicy.ts for an instant strength meter, with this module
as the one authoritative source of truth (the client-side check is
UX-only; a request that slips past it still gets rejected here).
"""

from __future__ import annotations

import re

MIN_LENGTH = 10
SPECIAL_CHARS = "!@#$%^&*()_+-=[]{}|;:,.<>?/~`"

# A short, deliberately small denylist of the most common passwords that
# would otherwise sail through every rule below (they're long enough, and
# some even mix case/digits). Not a substitute for a real breached-password
# API — just catching the most obvious cases without a network dependency.
_COMMON_PASSWORDS = {
    "password123",
    "password1234",
    "qwertyuiop12",
    "1234567890ab",
    "letmein12345",
    "iloveyou1234",
}


def validate_password(password: str, *, email: str | None = None) -> list[str]:
    """Returns a list of violated-rule messages; empty list means the
    password passes every rule. Never raises -- callers decide what to do
    with a non-empty list (e.g. auth.py turns it into a 422)."""
    violations: list[str] = []

    if len(password) < MIN_LENGTH:
        violations.append(f"Must be at least {MIN_LENGTH} characters long.")
    if not re.search(r"[a-z]", password):
        violations.append("Must include a lowercase letter.")
    if not re.search(r"[A-Z]", password):
        violations.append("Must include an uppercase letter.")
    if not re.search(r"\d", password):
        violations.append("Must include a number.")
    if not any(char in SPECIAL_CHARS for char in password):
        violations.append("Must include a special character (e.g. ! @ # $ % & *).")
    if password.lower() in _COMMON_PASSWORDS:
        violations.append("This password is too common — please choose another.")

    if email:
        local_part = email.split("@", 1)[0].lower()
        if local_part and local_part in password.lower():
            violations.append("Must not contain your email address.")

    return violations

File: app/utils/passwordPolicy.ts
/**
 * Mirrors api/app/services/password_policy.py's rules exactly, for an
 * instant client-side strength meter — same relationship as
 * discoverFilters.ts to discover.py. The backend re-checks every one of
 * these itself on signup; this file is UX only, never the source of truth.
 */

export const MIN_LENGTH = 10;
const SPECIAL_CHARS = "!@#$%^&*()_+-=[]{}|;:,.<>?/~`";

const COMMON_PASSWORDS = new Set([
  "password123",
  "password1234",
  "qwertyuiop12",
  "1234567890ab",
  "letmein12345",
  "iloveyou1234",
]);

export interface PasswordRuleResult {
  label: string;
  passed: boolean;
}

/** One entry per rule, in the same order the backend reports them, so a
 * live checklist can render check/x per rule as the user types. */
export function checkPasswordRules(password: string, email?: string): PasswordRuleResult[] {
  const rules: PasswordRuleResult[] = [
    { label: `At least ${MIN_LENGTH} characters`, passed: password.length >= MIN_LENGTH },
    { label: "A lowercase letter", passed: /[a-z]/.test(password) },
    { label: "An uppercase letter", passed: /[A-Z]/.test(password) },
    { label: "A number", passed: /\d/.test(password) },
    {
      label: "A special character (! @ # $ % & *)",
      passed: [...password].some((char) => SPECIAL_CHARS.includes(char)),
    },
  ];

  if (password.length > 0 && COMMON_PASSWORDS.has(password.toLowerCase())) {
    rules.push({ label: "Not a commonly used password", passed: false });
  }

  if (email) {
    const localPart = email.split("@")[0]?.toLowerCase();
    if (localPart) {
      rules.push({
        label: "Doesn't contain your email address",
        passed: !password.toLowerCase().includes(localPart),
      });
    }
  }

  return rules;
}

export function isPasswordValid(password: string, email?: string): boolean {
  return checkPasswordRules(password, email).every((rule) => rule.passed);
}
